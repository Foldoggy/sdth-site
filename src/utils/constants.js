export const MAPS_URL = "https://maps.google.com/maps?f=q&source=s_q&hl=en&geocode=&q="

export const NAV_HEIGHT = "88px"

export const pillarsInfo = Object.freeze({
  community: {
    color: "#90CC32"
  },
  education: {
    color: "#FFA91F"
  },
  inclusion: {
    color: "#0074BC"
  },
  talent: {
    color: "#2ABBF4"
  },
  innovation: {
    color: "#F25AA3"
  }
})
