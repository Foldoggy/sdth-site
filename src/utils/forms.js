export const notEmpty = (val) => val.length > 0

export const isValidEmail = (val) => /.@./.test(val)

export const usernameField = {
  name: "username",
  validate: notEmpty,
  errorMsg: "Please Enter Your Name"
}

export const emailField = {
  name: "email",
  validate: isValidEmail,
  errorMsg: "Email must be Valid"
}
