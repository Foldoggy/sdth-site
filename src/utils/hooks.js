import React, { useState } from "react"

export function useForm({ fields = [] }) {
  const form = { hasValidationErrors: false }
  form.fieldNames = fields.map(field => field.name)

  fields.forEach(field => {
    form[field.name] = useFormInput(field)
  })

  form.runValidations = () => {
    form.fieldNames.forEach(fieldName => {
      const field = form[fieldName]
      const isValid = field.validate(field.value)
      field.setIsValid(isValid)
      field.updateError(isValid)
    })

    form.hasValidationErrors = form.fieldNames.some(fieldName => {
      const field = form[fieldName]
      return !field.validate(field.value)
    })
  }

  return form
}

export function useFormInput({
  validate = () => true,
  errorMsg = "",
  initialValue = ""
}) {
  const [value, setValue] = useState(initialValue)
  const [isValid, setIsValid] = useState(true)
  const [error, setError] = useState("")

  const updateError = (nextIsValid) => setError(nextIsValid ? "" : errorMsg)

  const onChange = (e) => {
    const nextIsValid = validate(e.target.value)
    setIsValid(nextIsValid)
    setError(nextIsValid ? "" : errorMsg)
    setValue(e.target.value)
  }

  return {
    value,
    onChange,
    validate,
    isValid,
    setIsValid,
    error,
    updateError
  }
}

export function useCurator() {
  React.useEffect(() => {
    (function curatorIo() {
      const i = document.createElement("script")
      i.async = 1
      i.src = "https://cdn.curator.io/published/b46d667b-cd85-44f6-9f29-8991bf9f362e.js"
      const e = document.getElementsByTagName("script")[0]
      e.parentNode.insertBefore(i, e)
    }())
  }, [])
}
