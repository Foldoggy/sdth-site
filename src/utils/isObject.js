/**
 * Checks if value is an Object
 * @param {*} value
 * @returns {Boolean}
 */
const isObject = value => {
  if (!value) {
    return false
  }

  return value.constructor === Object
}

module.exports = isObject
