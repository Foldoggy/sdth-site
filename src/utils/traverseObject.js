const isObject = require("./isObject")

/**
 * This function will step through an object and apply a given function
 *
 * @param {Object} obj Object to be traversed
 * @param {Function} fn function to be applied to every key with "Description in it"
 * @returns {Object} New object with transformed keys
 */
function traverseObject(obj = {}, fn = (x) => x) {
  if (!isObject(obj)) {
    throw new Error(`first param must be an object. ${obj.contructor} given`)
  }

  const data = {}

  Object.keys(obj).forEach(key => {
    const value = obj[key]
    data[key] = value

    if (Array.isArray(value)) {
      data[key] = value.map(item => {
        if (isObject(item)) {
          return traverseObject(item, fn)
        }
        return item
      })
    }

    if (isObject(value)) {
      data[key] = traverseObject(value, fn)
    }

    if (typeof key === "string" && key.includes("Description")) {
      data[key] = fn(value)
    }
  })

  return data
}

module.exports = traverseObject
