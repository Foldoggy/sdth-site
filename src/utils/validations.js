export const notEmpty = (val) => val.length > 0

export const isValidEmail = (val) => /@./.test(val)
