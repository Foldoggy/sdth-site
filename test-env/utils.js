const { fireEvent } = require("react-testing-library")

export const MOCK_INVALID_EMAIL = "invalid@email"

export function asyncEvent() {
  return new Promise(resolve => {
    setTimeout(resolve, 0)
  })
}

export function fillInput(element, text) {
  fireEvent.change(element, { target: { value: text } })
}
