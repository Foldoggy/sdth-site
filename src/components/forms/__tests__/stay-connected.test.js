import React from "react"
import { fireEvent, render } from "react-testing-library"
import { fillInput } from "Test/utils"
import { library } from "@fortawesome/fontawesome-svg-core"
import MockAddToMailchimp from "gatsby-plugin-mailchimp"
import {
  faFacebookSquare,
  faLinkedin,
  faSlack,
  faTwitterSquare
} from "@fortawesome/free-brands-svg-icons"
import StayConnected from "../stay-connected"

library.add(
  faFacebookSquare,
  faLinkedin,
  faSlack,
  faTwitterSquare
)

jest.mock("gatsby-plugin-mailchimp", () => {
  return jest.fn(() => Promise.resolve({
    result: "success",
    msg: "Thanks for subscribing!"
  }))
})

test("matches its snapshot", () => {
  const { getByTestId } = render(<StayConnected />)
  expect(getByTestId("stay-connected")).toMatchSnapshot()
})

describe("Validation:", () => {
  test("prevents form submission unless name and email are valid", () => {
    const { getByLabelText, getByTestId } = render(<StayConnected />)
    const name = getByLabelText(/name/i)
    const comments = getByLabelText(/comments/i)
    const email = getByLabelText(/email/i)
    const subscribe = getByTestId("subscribe")

    // does NOT submit if name is blank
    fillInput(name, "")
    fillInput(email, "has@symbol")
    fillInput(comments, "a comment")
    fireEvent.click(subscribe)
    expect(MockAddToMailchimp).not.toHaveBeenCalled()

    // does NOT submit if email is invalid
    fillInput(name, "Name")
    fillInput(email, "noAtSymbol")
    fillInput(comments, "a comment")
    fireEvent.click(subscribe)
    expect(MockAddToMailchimp).not.toHaveBeenCalled()

    // DOES submit if name and email are valid
    fillInput(name, "Name")
    fillInput(email, "has@symbol")
    fillInput(comments, "")
    fireEvent.click(subscribe)
    expect(MockAddToMailchimp).toHaveBeenCalledTimes(1)
  })

  test("displays validation errors", () => {
    const { getByLabelText, getByTestId } = render(<StayConnected />)
    const name = getByLabelText(/name/i)
    const email = getByLabelText(/email/i)
    const subscribe = getByTestId("subscribe")

    // errors are NOT displayed initially
    expect(getByTestId("name-error")).toBeEmpty()
    expect(getByTestId("email-error")).toBeEmpty()

    // errors are displayed after submission attempt
    fireEvent.click(subscribe)
    expect(getByTestId("name-error")).not.toBeEmpty()
    expect(getByTestId("email-error")).not.toBeEmpty()

    // errors disappear when fields become valid
    fillInput(name, "Name")
    expect(getByTestId("name-error")).toBeEmpty()
    fillInput(email, "has@symbol")
    expect(getByTestId("email-error")).toBeEmpty()

    // errors reappear when fields become invalid
    fillInput(name, "")
    expect(getByTestId("name-error")).not.toBeEmpty()
    fillInput(email, "noAtSymbol")
    expect(getByTestId("email-error")).not.toBeEmpty()
  })
})

describe("Success Message:", () => {
  xtest("displays success message if subscribed", () => {

  })

  xtest("does NOT display success message if validation errors", () => {

  })
})
