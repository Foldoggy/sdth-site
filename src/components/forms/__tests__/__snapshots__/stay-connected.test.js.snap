// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches its snapshot 1`] = `
<div
  class="sc-bwzfXH jSEfmW"
  data-testid="stay-connected"
>
  <form
    class="sc-htpNat bZbnzF"
    method="post"
  >
    <h2
      class="bigScreen sc-bxivhb jDwgEc"
    >
      Stay Connected
    </h2>
    <div>
      <label
        for="name"
      >
        Name:
        <input
          id="name"
          value=""
        />
      </label>
      <div
        data-testid="name-error"
        style="color: red; height: 2rem;"
      />
    </div>
    <div>
      <label
        for="email"
      >
        Email:
        <input
          id="email"
          type="email"
          value=""
        />
      </label>
      <div
        data-testid="name-error"
        style="color: red; height: 2rem;"
      />
    </div>
    <div>
      <label
        for="comments"
      >
        Comments:
        <textarea
          id="comments"
        />
      </label>
    </div>
    <button
      data-testid="subscribe"
      type="submit"
    >
      Join the Movement
    </button>
  </form>
  <div
    class="sc-ifAKCX cejGdI"
  >
    <p>
      Want to be a conduit for change? Join the movement and connect with liked-minded individuals looking to make a difference redefining the San Diego tech scene.
    </p>
    <a
      class="sc-bdVaJa erbOLL"
      href="https://join.slack.com/t/sandiegotechhub/shared_invite/enQtNTI1MDA2NjQyNDcwLTRhYmFhOGZlNzQyZWQ0NmJjMTEzNGE1YjI1NTJmY2RhZjVmYjBjNDAyYmI4MDZkNTM4MzMwM2JmYWQzOGVkYjY"
      rel="noopener noreferrer"
      target="_blank"
    >
      <button
        type="submit"
      >
        <svg
          aria-hidden="true"
          class="svg-inline--fa fa-slack fa-w-14 fa-sm "
          data-icon="slack"
          data-prefix="fab"
          role="img"
          viewBox="0 0 448 512"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M244.2 217.5l19.3 57.7-59.8 20-19.3-57.7 59.8-20zm41.4 243.7C131.6 507.4 65 471.6 18.8 317.6S8.4 97 162.4 50.8C316.4 4.6 383 40.4 429.2 194.4c46.2 154 10.4 220.6-143.6 266.8zM366.2 265c-3.9-12.2-17.2-18.6-29.4-14.7l-29 9.7-19.3-57.7 29-9.7c12.2-3.9 18.6-17.2 14.7-29.4-3.9-12.2-17.2-18.6-29.4-14.7l-29 9.7-10-30.1c-3.9-12.2-17.2-18.6-29.4-14.7-12.2 3.9-18.6 17.2-14.7 29.4l10 30.1-59.8 20.1-10-30.1c-3.9-12.2-17.2-18.6-29.4-14.7-12.2 3.9-18.6 17.2-14.7 29.4l10 30.1-29 9.7c-12.2 3.9-18.6 17.2-14.7 29.4 3.2 9.3 12.2 15.4 21.5 15.8 4.3.6 7.7-1 36.9-10.7l19.3 57.7-29 9.7c-12.2 3.9-18.6 17.2-14.7 29.4 3.2 9.3 12.2 15.4 21.5 15.8 4.3.6 7.7-1 36.9-10.7l10 30.1c3.7 10.8 15.8 18.6 29.4 14.7 12.2-3.9 18.6-17.2 14.7-29.4l-10-30.1 59.8-20.1 10 30.1c3.7 10.8 15.8 18.6 29.4 14.7 12.2-3.9 18.6-17.2 14.7-29.4l-10-30.1 29-9.7c12.2-4.2 18.6-17.5 14.7-29.6z"
            fill="currentColor"
          />
        </svg>
        <span
          style="margin-left: 1rem;"
        >
          Join our Slack 
          <span
            class="hidden-on-mobile"
          >
            Community
          </span>
        </span>
      </button>
    </a>
    <aside
      style="text-align: center; padding: 1rem; margin-top: 1rem;"
    >
      <h4
        style="font-size: 1.4rem;"
      >
        Like and follow San Diego Tech Hub
      </h4>
      <div
        style="display: flex; justify-content: space-evenly; width: 75%; margin: auto;"
      >
        <a
          class="sc-bdVaJa erbOLL"
          href="https://www.linkedin.com/company/san-diego-tech-hub/"
          rel="noopener noreferrer"
          target="_blank"
        >
          <svg
            aria-hidden="true"
            class="svg-inline--fa fa-linkedin fa-w-14 fa-2x "
            color="#0077B5"
            data-icon="linkedin"
            data-prefix="fab"
            role="img"
            viewBox="0 0 448 512"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M416 32H31.9C14.3 32 0 46.5 0 64.3v383.4C0 465.5 14.3 480 31.9 480H416c17.6 0 32-14.5 32-32.3V64.3c0-17.8-14.4-32.3-32-32.3zM135.4 416H69V202.2h66.5V416zm-33.2-243c-21.3 0-38.5-17.3-38.5-38.5S80.9 96 102.2 96c21.2 0 38.5 17.3 38.5 38.5 0 21.3-17.2 38.5-38.5 38.5zm282.1 243h-66.4V312c0-24.8-.5-56.7-34.5-56.7-34.6 0-39.9 27-39.9 54.9V416h-66.4V202.2h63.7v29.2h.9c8.9-16.8 30.6-34.5 62.9-34.5 67.2 0 79.7 44.3 79.7 101.9V416z"
              fill="currentColor"
            />
          </svg>
        </a>
        <a
          class="sc-bdVaJa erbOLL"
          href="https://www.facebook.com/SanDiegoTechHub/"
          rel="noopener noreferrer"
          target="_blank"
        >
          <svg
            aria-hidden="true"
            class="svg-inline--fa fa-facebook-square fa-w-14 fa-2x "
            color="#3b5998"
            data-icon="facebook-square"
            data-prefix="fab"
            role="img"
            viewBox="0 0 448 512"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M448 80v352c0 26.5-21.5 48-48 48h-85.3V302.8h60.6l8.7-67.6h-69.3V192c0-19.6 5.4-32.9 33.5-32.9H384V98.7c-6.2-.8-27.4-2.7-52.2-2.7-51.6 0-87 31.5-87 89.4v49.9H184v67.6h60.9V480H48c-26.5 0-48-21.5-48-48V80c0-26.5 21.5-48 48-48h352c26.5 0 48 21.5 48 48z"
              fill="currentColor"
            />
          </svg>
        </a>
        <a
          class="sc-bdVaJa erbOLL"
          href="https://twitter.com/SanDiegoTechHub"
          rel="noopener noreferrer"
          target="_blank"
        >
          <svg
            aria-hidden="true"
            class="svg-inline--fa fa-twitter-square fa-w-14 fa-2x "
            color="#00a1f8"
            data-icon="twitter-square"
            data-prefix="fab"
            role="img"
            viewBox="0 0 448 512"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M400 32H48C21.5 32 0 53.5 0 80v352c0 26.5 21.5 48 48 48h352c26.5 0 48-21.5 48-48V80c0-26.5-21.5-48-48-48zm-48.9 158.8c.2 2.8.2 5.7.2 8.5 0 86.7-66 186.6-186.6 186.6-37.2 0-71.7-10.8-100.7-29.4 5.3.6 10.4.8 15.8.8 30.7 0 58.9-10.4 81.4-28-28.8-.6-53-19.5-61.3-45.5 10.1 1.5 19.2 1.5 29.6-1.2-30-6.1-52.5-32.5-52.5-64.4v-.8c8.7 4.9 18.9 7.9 29.6 8.3a65.447 65.447 0 0 1-29.2-54.6c0-12.2 3.2-23.4 8.9-33.1 32.3 39.8 80.8 65.8 135.2 68.6-9.3-44.5 24-80.6 64-80.6 18.9 0 35.9 7.9 47.9 20.7 14.8-2.8 29-8.3 41.6-15.8-4.9 15.2-15.2 28-28.8 36.1 13.2-1.4 26-5.1 37.8-10.2-8.9 13.1-20.1 24.7-32.9 34z"
              fill="currentColor"
            />
          </svg>
        </a>
      </div>
    </aside>
  </div>
  <h2
    class="smallScreen"
  >
    Stay Connected
  </h2>
</div>
`;
